tests/zfs-tests/tests/functional/zvol/zvol_stress/zvol_stress.ksh

```diff
@@ -0,0 +1,161 @@
+#
+# CDDL HEADER END
+#
+# Copyright (c) 2022 by Lawrence Livermore National Security, LLC.
+
+. $STF_SUITE/include/libtest.shlib
+. $STF_SUITE/tests/functional/reservation/reservation.shlib
+. $STF_SUITE/tests/functional/zvol/zvol_common.shlib
+
+#
+# DESCRIPTION:
+# Stress test multithreaded transfers to multiple zvols.  Also verify
+# zvol errors show up in zpool status.
+#
+# STRATEGY:
+#
+# For both the normal submit_bio() codepath and the blk-mq codepath, do
+# the following:
+#
+# 1. Create one zvol per CPU
+# 2. In parallel, spawn an fio "write and verify" for each zvol
+# 3. Inject write errors
+# 4. Write to one of the zvols with dd and verify the errors
+#
+
+verify_runnable "global"
+
+num_zvols=$(get_num_cpus)
+
+# If we were making one big zvol from all the pool space, it would
+# be this big:
+biggest_zvol_size_possible=$(largest_volsize_from_pool $TESTPOOL)
+
+# Crude calculation: take the biggest zvol size we could possibly
+# create, knock 10% off it (for overhead) and divide by the number
+# of ZVOLs we want to make.
+each_zvol_size=$(( floor($biggest_zvol_size_possible * 0.9) / $num_zvols ))
+
+typeset tmpdir="$(mktemp -d zvol_stress_fio_state.XXXXXX)"
+
+function create_zvols
+{
+	log_note "Creating $num_zvols zvols that are ${each_zvol_size}B each"
+	for i in $(seq $num_zvols) ; do
+		log_must zfs create -V $each_zvol_size $TESTPOOL/testvol$i
+		block_device_wait "$ZVOL_DEVDIR/$TESTPOOL/testvol$i"
+	done
+}
+
+function destroy_zvols
+{
+	for i in $(seq $num_zvols) ; do
+		log_must_busy zfs destroy $TESTPOOL/testvol$i
+	done
+}
+
+function do_zvol_stress
+{
+	# Write 10% of each zvol, or 50MB, whichever is less
+	zvol_write_size=$((each_zvol_size / 10))
+	if [ $zvol_write_size -gt $((50 * 1048576)) ] ; then
+		zvol_write_size=$((50 * 1048576))
+	fi
+	zvol_write_size_mb=$(($zvol_write_size / 1048576))
+
+	if is_linux ; then
+		engine=libaio
+	else
+		engine=psync
+	fi
+
+	# Spawn off one fio per zvol in parallel
+	pids=""
+	for i in $(seq $num_zvols) ; do
+		# Spawn one fio per zvol as its own process
+		fio --ioengine=$engine --name=zvol_stress$i --direct=0 \
+			--filename="$ZVOL_DEVDIR/$TESTPOOL/testvol$i" --bs=1048576 \
+			--iodepth=10 --readwrite=randwrite --size=${zvol_write_size} \
+			--verify_async=2 --numjobs=1 --verify=sha1 \
+			--verify_fatal=1 \
+			--continue_on_error=none \
+			--error_dump=1 \
+			--exitall_on_error \
+			--aux-path="$tmpdir" --do_verify=1 &
+		pids="$pids $!"
+	done
+
+	# Wait for all the spawned fios to finish and look for errors
+	fail=""
+	i=0
+	for pid in $pids ; do
+		log_note "$s waiting on $pid"
+		if ! wait $pid ; then
+			log_fail "fio error on $TESTPOOL/testvol$i"
+		fi
+		i=$(($i + 1))
+	done
+}
+
+function cleanup
+{
+	log_must zinject -c all
+	log_must zpool clear $TESTPOOL
+	destroy_zvols
+	set_blk_mq 0
+
+	# Remove all fio's leftover state files
+	if [ -n "$tmpdir" ] ; then
+		rm -f "$tmpdir"/*.state
+		rmdir "$tmpdir"
+	fi
```

Comment:
is this any different from `rm -fd "$tmpdir"/*.state "$tmpdir"`?

Reply:
Yep, that will work better.  I'll fix